AnkiHubSoftware/ankihub_addon
# fix: Add migration to re-fetch `DeckMedia`

This PR adds a migration which causes the add-on to re-fetch all `DeckMedia` for subscribed decks if `DeckMedia` was fetched before version 15.0 of the AnkiHub API. This is needed because the `DeckMediaListAPIView` previously skipped some entries.

## Context
I made a [post](https://theanking.slack.com/archives/C01E9EJ1W2E/p1694185448350989) about this on Slack, here is the message:
> I temporarily disabled the `use_deck_mediafeature` flag, because Heitor and I noticed yesterday that there could a problem with the cursor pagination used in the `DeckMediaListAPIView` and I confirmed the problem today. The cursor pagination can skip `DeckMedia` entries if they have the same modified timestamp. Missing `DeckMedia` entries could lead to media not getting downloaded to Anki. We haven't seen any complaints yet.We already had the same problem with other views that use DRF's cursor pagination.
I created a PR that fixes the problem: [Fix: Use other cursor pagination implementation for DeckMediaListAPIView](https://github.com/ankipalace/ankihub/pull/1708). Here's the [notion task](https://www.notion.so/ankihub/AnkiHub-Planning-7d2bdb8e15034c559bef129adbe163b7?p=4f853286bce24337ad9ae22c7b657375&pm=s).
Once this is reviewed we can apply the same change to the other views that use cursor pagination to make the code cleaner (the other views use a workaround to prevent the issue).Until this is done, the add-on will use the old media sync process, as it did before the release of the s3/DeckMedia project.
When we turn the feature flag on again, we will need to add a migration to the add-on so that it re-fetches all `DeckMedia` object to make sure that none are missing.All in all, it's not too bad, we just need to fix the cursor pagination and can then re-enable the feature flag soon after.

## Related issues

Notion task: [fix: Add migration to add-on to re-fetch DeckMedia](https://www.notion.so/ankihub/AnkiHub-Planning-7d2bdb8e15034c559bef129adbe163b7?p=493be4e38ff442bf8c56d3b8eeb4ae91&pm=s)

These two PRs bump the api version to 15.0 for the add-on and web app. They need to be merged before the changes introduce in this PR are released:
- #745
- https://github.com/ankipalace/ankihub/pull/1715


## Proposed changes
Changes:
- Added an `api_version_on_last_sync` field to the private config of the add-on:
https://github.com/ankipalace/ankihub_addon/blob/8c111bace622a679bb552719a9baa7f371353176/ankihub/settings.py#L120-L121
This field is updated when an AnkiHub sync finishes.

- Added an migration to `private_config_migrations.py`. If `api_version_on_last_sync` is below 15.0, the `latest_media_update` timestamp is reset to `None` for all decks. (`DeckMediaListAPIView` will be fixed in version 15.0) This causes the add-on to re-fetch `DeckMedia` for all decks.

Once the add-on fetches `DeckMedia` on api version 15.0 or above, the migration won't be applied anymore.

By using `api_verson_on_last_sync` we can ensure that migrations are applied only once when an API version changes. This will probably be useful for future migrations as well.

<a href="https://gitpod.io/#https://github.com/ankipalace/ankihub_addon/pull/744"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

## How to reproduce
- This assumes you are syncing with API version 14.0. 
- Open the `.private_config.json` file for the currently used Anki profile.
- Sync with a deck that has media. Observer that the `latest_media_update` value for the deck  in the json was updated to a timestamp.
- Restart Anki. Observe that `latest_media_update` was set to `null` in the json (by the migration).
- Sync with AnkiHub. Observe that `latest_media_update` is set to a timestamp again.

When you sync with API version 15.0, `latest_media_update` won't be set to `null` on every Anki start, because the migration won't be applied.

See this comment for more details on why using api_version_on_last_sync to determine which migrations to run is effective:
https://github.com/ankipalace/ankihub_addon/blob/0f98e09025b1d8e6623c192df89158d13638c08e/ankihub/private_config_migrations.py#L16-L20
